packages/jest-jasmine2/src/jasmineAsyncInstall.ts

```diff
@@ -17,26 +17,30 @@ import throat from 'throat';
 import isError from './isError';
 import type {Jasmine} from './types';
 import type Spec from './jasmine/Spec';
-import type {QueueableFn} from './queueRunner';
+import type {DoneFn, QueueableFn} from './queueRunner';
 
-interface DoneFn {
-  (): void;
-  fail: (error: Error) => void;
-}
-
-function isPromise(obj: any) {
+function isPromise(obj: any): obj is PromiseLike<unknown> {
   return obj && typeof obj.then === 'function';
 }
 
+const doneFnNoop = () => {};
+
+doneFnNoop.fail = () => {};
+
 function promisifyLifeCycleFunction(
-  originalFn: Function,
+  originalFn: (beforeAllFunction: QueueableFn['fn'], timeout?: number) => void,
   env: Jasmine['currentEnv_'],
 ) {
   return function <T>(
-    fn: Function | (() => Promise<T>) | GeneratorFunction | undefined,
+    fn:
+      | ((done: DoneFn) => void | PromiseLike<T>)
```

Comment:
It should be an error to both take a done function _and_ return a promise

Reply:
This will cause some major type problems, as everywhere else is expecting only `(done: DoneFn) => void`.

For the record, by unioning the functions TS can't support both of them when doing `.call`, so it favors `(done: Done) => void`, meaning that `returnValue.then` is invalid as it has type `void`.

This can be resolved by using an interface:

```
interface OneOrTheOther {
  (done: DoneFn): void;
  (): PromiseLike<void>;
}
```

This then takes us back to where we were before, but this time it's the opposite: TS is complaining `asyncJestTest` doesn't fit because it's of type `(done: DoneFn) => void` instead of `OneOrTheOther` 😬

Reply:
😞